yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -225,12 +240,89 @@ private[yarn] class YarnAllocator(
+
+      if (updatedLocalityAwareContainerNum == 0) {
+        // If the current allocated executor can satisfy all the locality preferred tasks,
+        // allocate the new container with no locality preference
+        requiredLocalityFreeContainerNum = missing
+      } else {
+        if (updatedLocalityAwareContainerNum >= missing) {
+          // If newly requested containers cannot satisfy the locality preferred tasks,
+          // allocate all the new container with locality preference
+          requiredLocalityAwareContainerNum = missing
+        } else {
+          // If part of newly requested can satisfy the locality preferred tasks, allocate part of
+          // the containers with locality preference, and another part with no locality preference
+          requiredLocalityAwareContainerNum = updatedLocalityAwareContainerNum
+          requiredLocalityFreeContainerNum = missing - updatedLocalityAwareContainerNum
+        }
+      }
+
+      if (requiredLocalityFreeContainerNum > 0) {
+        for (i <- 0 until requiredLocalityFreeContainerNum) {
+          val request = createContainerRequest(resource, null, null)
+          amClient.addContainerRequest(request)
+          val nodes = request.getNodes
+          val hostStr = if (nodes == null || nodes.isEmpty) "Any" else nodes.last
+          logInfo(s"Container request (host: $hostStr, capability: $resource)")
+        }
+      }
+
+      if (requiredLocalityAwareContainerNum > 0) {
+        val largestRatio = expectedLocalityToContainerNum.values.max
+        // Round the ratio of preferred locality to the number of locality required container
+        // number, which is used for locality preferred host calculating.
+        var preferredLocalityRatio = expectedLocalityToContainerNum.mapValues { ratio =>
+          val adjustedRatio = ratio.toDouble * requiredLocalityAwareContainerNum / largestRatio
+          adjustedRatio.ceil.toInt
+        }
+
+        for (i <- 0 until requiredLocalityAwareContainerNum) {
+          // Only filter out the ratio which is larger than 0, which means the current host can
+          // still be allocated with new container request.
+          val hosts = preferredLocalityRatio.filter(_._2 > 0).keys.toArray
+          val racks = hosts.map(h => RackResolver.resolve(conf, h).getNetworkLocation).toSet
+          val request = createContainerRequest(resource, hosts, racks.toArray)
+          amClient.addContainerRequest(request)
+          val nodes = request.getNodes
+          val hostStr = if (nodes == null || nodes.isEmpty) "Any" else nodes.last
+          logInfo(s"Container request (host: $hostStr, capability: $resource)")
+
+          // Each time when the host is used, subtract 1. When the current ratio is 0,
+          // which means all the required ratio is satisfied, this host will not allocated again.
+          // Details can be seen in the SPARK-4352.
+          preferredLocalityRatio = preferredLocalityRatio.mapValues(i => i - 1)
```

Comment:
Thanks a lot @squito , greatly appreciate your comments. I'm still working on this, my first target is to prove the algorithm that it could be worked, and then I will refactor the code to make it more clear. Sorry about few comments on the code, thanks a lot for your comments :smiley: .